WIP - Window Covering

newcomertv opened the pull request:
An initial draft of a Window Covering Cluster with example Device.

Currently only mandatory attributes exist. Device is commissionable to Homekit but gets displayed as unresponsive.

Not sure how best to define the Maps, hoping for feedback on that.
The Cluster implementation needs to vary depending on the features of the device.
I propose setting the attributes inside Option enums and implementing a general cluster checking for the current feature map.

Won't continue work on this until 22.08.2023, contributions are welcome to flesh this out.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/matter-rs?pullRequest=86) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/matter-rs?pullRequest=86) before we can accept your contribution.<br/><hr/>**Nikolas von Lonski** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/matter-rs?pullRequest=86) it.</sub>

ivmarkov commented on rs-matter/src/data_model/cluster_window_covering.rs:
```diff
@@ -0,0 +1,658 @@
+                        codec.encode(writer, self.end_product_type.get())
+                    }
+                    Attributes::Mode(codec) => codec.encode(writer, self.mode.get()),
+                    _ => todo!("read {:?}", attr.attr_id),
```
Would that crash on any other attribute and is this the intended behavior?

newcomertv replied:
If the attribute is currently not supported I'd expect the function to return such an error

ivmarkov commented on rs-matter/src/data_model/cluster_window_covering.rs:
```diff
@@ -0,0 +1,658 @@
+
+        match attr.attr_id.try_into()? {
+            Attributes::Type(codec) => self.set_type(codec.decode(data)?),
+            _ => todo!("write {:?}", attr.attr_id),
```
Ditto. Is any other attribute not expected here or is this unfinished code-path?

newcomertv replied:
Unfinished code path, if an attribute is unexpected an error should be propagated up the call stack

ivmarkov commented on rs-matter/src/data_model/cluster_window_covering.rs:
```diff
@@ -0,0 +1,658 @@
+        _data: &TLVElement,
+        _encoder: CmdDataEncoder,
+    ) -> Result<(), Error> {
+        match cmd.cmd_id.try_into()? {
```
Ditto. Would that crash?

newcomertv replied:
Ditto. Errors should be propagated up the call stack. Do you disagree how it is handled here?

ivmarkov replied:
`todo!` panics. It does not propagate errors up the call stack. It will essentially crash your app.

ivmarkov commented:
@newcomertv There are some Clippy issues that you might need to address as well.

newcomertv commented:
State : Cluster can get commissioned and receives commands however state remains as "No response received" within Apple HomeKit. 
Fixed clippy warnings and formatting.
Clarified incomplete sections.
Added a builder pattern skeleton as this cluster accepts various feature combinations.
